File: 14.UnionFindSet并查集/leetcode_1722_minimumHammingDistance.py
from typing import List
from collections import defaultdict

class UFS:
    def __init__(self, n):
        self.pa = list(range(n))
    
    def find(self, x):
        if self.pa[x] != x:
            self.pa[x] = self.find(self.pa[x])
        return self.pa[x]
    
    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        if px < py:
            self.pa[py] = px
        else:
            self.pa[px] = py

class Solution:
    def minimumHammingDistance(self, source: List[int], target: List[int], allowedSwaps: List[List[int]]) -> int:
        n = len(source)
        ufs = UFS(n)
        for u,v in allowedSwaps:
            ufs.union(u, v)
        
        groups = defaultdict(list)
        for i in range(n):
            groups[ufs.find(i)].append(i)
            
        def calc(pos):
            cnt = defaultdict(int)
            for i in pos:
                cnt[source[i]] += 1
                cnt[target[i]] -= 1
            ans = 0
            for v in cnt.values():
                ans += abs(v)
            return ans//2
        
        res = 0
        for pos in groups.values():
            res += calc(pos)
        return res

if __name__ == "__main__":
    s = Solution()
    source = [1,2,3,4]
    target = [2,1,4,5]
    allowedSwaps = [[0,1],[2,3]]
    print(s.minimumHammingDistance(source, target, allowedSwaps))

File: 01.DFS递归/leetcode_1722_minimumHammingDistance.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
private:
    vector<int> visited, Source, Target;
    vector<vector<int>> g;
    unordered_map<int, int> cnt;

    void dfs(int x){
        cnt[Source[x]]++;
        cnt[Target[x]]--;
        visited[x] = 1;
        for (int y : g[x]){
            if (visited[y]){
                continue;
            }
            dfs(y);
        }
    }
public:
    int minimumHammingDistance(vector<int>& source, vector<int>& target, vector<vector<int>>& allowedSwaps) {
        int n = source.size();
        Source = source;
        Target = target;
        g.resize(n);
        visited.resize(n, 0);
        for (auto& aS : allowedSwaps){
            g[aS[0]].push_back(aS[1]);
            g[aS[1]].push_back(aS[0]);
        }

        int res = 0;
        for (int i=0; i<n; i++){
            if (visited[i] == 0){
                cnt.clear();
                dfs(i);
                for (auto& kv : cnt){
                    res += abs(kv.second);
                }
            }
        }
        return res/2;
    }
};

int main(){
    Solution s;
    vector<int> source = {1,2,3,4}, target = {2,1,4,5};
    vector<vector<int>> allowedSwaps = {{0,1},{2,3}};
    cout << s.minimumHammingDistance(source, target, allowedSwaps) << endl;
}
